Editorial changes to measurement discussion (#248)
ietf-rats-wg/draft-ietf-rats-corim: draft-ietf-rats-corim.md

A Verifier needs to reconcile its various inputs, with CoRIM being one of them.
In addition to the external CoRIM documents, the Verifier is expected to create an internal representation for each input and map each external representation to an internal one.
By using the internal representation, the Verifier processes inputs as if they are part of a conversation, keeping track of who said what.

Effectively, Attesters, Reference Value Providers, Endorsers, Verifier Owners, Relying Parties, and even the Verifier potentially all contribute to the conversation.
These individual principals of the conversation may have their identity rooted back to some *authority* farther up the certificate chain than the signing key itself, such as a digital certificate authority intermediate or root key.

> The origin of contributions/inputs to the conversation are tracked at the granularity of authority, so CoRIM authors can predicate their reference measurements or endorsements based on existing properties rooted from the same authority.

Reviewer: This might require en example to be less abstract?

Author: Added. Resolve if it's okay.

Reviewer 2: @Dionna This is aintroductory text, making some high level statements. 
- This is way too a technical details in the introductory sections
- Specific items like authority for entities like Relying Party are not in scope of this document.

> CoRIM inputs to the conversation include measurements within a specific Environment.

Reviewer: you are introducing "the conversation" here. Is there some similar wording that already exists in the I-D to use instead. Otherwise, we have to clarify a bit of the scope, involved actors, and subject of the conversation, I think

Author: I'm not introducing it here. It's introduced in text immediately prior about all parties that contribute to "the conversation" with authorities tracking "who said what".

> A measurement never appears outside the scope of an associated Environment, so the Environment-Claim Tuple (ECT) is a data structure used to encapsulate all scope context for a given set of measurements.

Reviewer: Maybe: 
```suggestion
A specific set of measuremenst always originates from a single identifiable and associated Environment. Hence. an Environment-Claim Tuple (ECT) is a data structure used to encapsulate that scope context for a given set of measurements.
```

Author: This suggestion is saying much more than I'm saying, since I don't insist that each specific set of measurements has a single environment. Indeed, sets of measurements don't necessarily have a globally unique identifier for distinguishing them across individual measurements. I can have the same card on two different sockets and have the same measurement set, but a different environment due to my assignment of card to group.

Perhaps "specific set" is too ambiguous. Let's go with "a specific occurrence of a set", since the occurrence refers to where it is in the CoRIM.

The authority for an ECT from a CoRIM is the CoRIM issuer.

> The collective content of an ECT may be referred to as a property, claim, or assertion when the content need not be presented in the ECT data structure.

Reviewer: ```suggestion
When the content of an a ECT data structure is presented to roles external to the Verifier, it MAY be referred to via different terms, such as property, claim, or assertion. Whenever an ECT is consumed by a role external to the Verifer, the consuming party MUST always attribute provenance and the ECT's intended use.
```

Author: What is "intended use"? I'm not sure how to comply with that. My code that touches it could be my intention because I'm a hot shot with "self-documenting code". Is that enough?

I thought ECT is a representation of an utterance in the conversation, with all scope-closing elements provided.

Important: The Verifier's objective is to produce a list of properties that describe the Attester's presumed actual state.

> Each producer of corresponding RATS Conceptual Messages can assert claims about an Attester's actual or allowed state.

Reviewer: Are you also implying that Evidence can assert claims about allowed state?

Author: Conceptual messages aren't all evidence. I thought there was already a cm-ind allocated for policy?

> Each producer of corresponding RATS Conceptual Messages can assert claims about an Attester's actual or allowed state.

Author: Yes, fixed.

* `group` (index 2): identifier for a group of instances, e.g., if an
  anonymization scheme is used.

An `environment-map` contains at least one of `class`, `instance`, or `group`.

> The more elements in the `environment-map`, the narrower the scope, since more elements must be in common to match.

Reviewer 2: ```suggestion
```

##### Measurements

> A measurement can be directly compared with attestation evidence, or it can be a quality that is not directly comparable with evidence.

Reviewer: I cannot really follow "[Evidence] can be a quality that is not directly comparable with evidence". Could you maybe provide an example so it becomes more clear?

Reviewer: In general, I think that Evidence is composed of "Measurements" and there might be a bigger disconnect here that I cannot follow.

Author: I don't think that an assigned name or serial number is always provided by an attesting environment's measurement mechanism. That information is added metadata provided the reference values are as expected.

pseudocode:

```
reference-values-triple [E, M[integrity-registers: X]]
conditional-endorsement-triple [
  conds: [E, [integrity-registers: X authorities: [me]]]
  [E, M'[name: "Product: Enterprise Edition", svn: 5]]
]
```

I'm not really sure how to express associated metadata in a less-clunky way, unless this conditionality is what is the expected semantics of the non-conditional endorsement-triple. It's unclear from the text https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/179#issuecomment-2184116580

> The scope of a mroperty is defined by the `environment-map` and optional `mkey` it appears with.

Reviewer: ```suggestion
The scope of a property is defined by the `environment-map` and optional `mkey` it appears with.
```

reconfiguration regions, etc. Measurements comprise raw values, digests, or
status information.
A measurement can be directly compared with attestation evidence, or it can be a quality that is not directly comparable with evidence.
The scope of a mroperty is defined by the `environment-map` and optional `mkey` it appears with.

An environment has one or more measurable elements. Each element can have a
dedicated measurement or multiple elements could be combined into a single
measurement. Measurements can have class, instance or group scope.  This is

> typically determined by the triple's environment.

Author: I find lowercase "environment" difficult to map to a single concept. Here's my understanding. An `environment-map` does not contain measurements. An Attesting Environment has measurable elements. A Target Environment has reference values for elements.

I reword or reformat the rest of these sentences below. I don't like "typically" here since measurements always are within the scope on an environment-map.

> typically determined by the triple's environment.

Reviewer 3: An environment (aka Target Environment) is the thing that is being measured by an Attesting Environment. The name of the environment is expressed as key-value pairs (which is what environment-map and stateful-environment-record do). An AE collects measurements about the TE and those measurements are expressed as key-value pairs. The measurement-value-map structure does this. 

We can change "environment" to use TE if lower case environment is confusing.
I'm OK with removing "typically".

> Measurements can be of a variety of things including software, firmware, configuration files, read-only memory, fuses, IO ring configuration, partial reconfiguration regions, etc. Measurements comprise raw values, digests, or status information.

Reviewer: ```suggestion
Measurements can be of a variety of things including measurement values about software, firmware, configuration files, read-only memory, fuses, IO ring configuration, partial reconfiguration regions, etc. Measurements comprise raw values, digests, or status information.
```

> Each element can have a dedicated measurement/endorsement value or multiple elements could be combined into a single measurement.

Reviewer: This might benefit from another example.

Author: Added. Resolve if it's okay.

> The more elements in the `environment-map`, the narrower the scope, since more elements must be in common to match.

Author: Moved.

> Measurements paired with an `environment-map` that contains only a class may be referred to as class mroperties, and similarly with instance and group measurements for respective elements of the `environment-map`.

Reviewer: ```suggestion
Measurements paired with an `environment-map` that contains only a class may be referred to as class properties, and similarly with instance and group measurements for respective elements of the `environment-map`.
```

A measurement in a `reference-triple-record` is called a Reference Value.

> A measurement in another triple is called an Endorsement.

Reviewer: Another triple?

Author: Every non-reference-value-triple that contains measurement-map is some flavor of endorsement, so that's the most concise way I could state that. I'm still unclear if there's a conditional relationship between reference triples and the rest of the triples in the same CoMID.

Reviewer 3: >I'm still unclear if there's a conditional relationship between reference triples and the rest of the triples 

If they're related, the relationship is defined by the 'condition' portion of the RV triple. In the `reference-triples` case, both the `environment-map` and `measurement-map` are the condition. If there is a match, the "corroborated" values (the actual state representation) is added to the ACS under the RVP authority. If RVs are already expressed as actual state representations, then you can say "corroboration" is passthrough and the RVPs are copied into the ACS (differing only by authority from what was matched).

There is an ongoing discussion whether RVs should be used to "corroborate" ACS values that originated from Endorsers vs. Attesters. Both are representations of **actual state**. So, its technically possible. The use case for allowing it is when there are claims that can't reasonably be asserted by an AE that a vendor of that Attester wishes to supply reference values. A contrived example is if a CC certifying lab reports that an Attester meets EAL4+, but the Attester vendor wants to restrict the upper bound (for some reason), then the attester vendor could create a RV for a range, say EAL2+ to EAL4. The RV condition applies to claims that only an endorser would make. But the Verifier can easily process them just like RVs written to match Evidence.

The concern is a constrained verifier might not have resources necessary to handle a large set of inputs and one way to reduce them is to disallow processing of RVs aimed at matching Endorsements.

Reviewer 3: > Another triple?

It would be clearer to mention a specific endorsement triple such as `endorsed-triples`

###### Measurement Keys {#sec-comid-mkey}

> A measurement-associated (sub-)environment is at the granularity of a single element, such as the second enumerated network interface card.

Author: I would like to, but given the existing use of "measured-element-type-choice" to refer to those logically composed objects, I'd like to know what word we want to use that's not the "element" from the CDDL spec. Do we want to say measured-object?

> A measurement-associated (sub-)environment is at the granularity of a single element, such as the second enumerated network interface card.

Reviewer 3: I suppose everyone has different intuition around what words to use to describe compositions of things. Using RFC9334 as a guide, it uses environment (TE/AE) and "sub-Attester" which is another way of saying some sort of construction of TEs and AEs, but it didn't use "sub-TE" or "sub-AE" and some would find this awkward. 

The current wording reflects the idea that a TE can be broken into finer granularity sub-parts and the English dictionary definition of "element" is a word that isn't too overloaded.
If a lay person described a computing environment as something that consisted of several sub-elements it wouldn't be too far from reality while still being abstract enough to fit a wide variety of implementations.

> A measurement-associated (sub-)environment is at the granularity of a single element, such as the second enumerated network interface card.

Author: I say (sub-)enviroment consistently with the existing text for `mkey`. Notion of it being a sub-environment is that it is further limiting the scope of the subject named by the `environment-map`, but is not itself an element of `environment-map`.

I changed from "element" to "object-id" since I read that you wanted element to consistently mean the same thing, i.e., the meaning from CDDL. Are you suggesting that "element" is not too overloaded as an English term, so we _don't_ have to use "element" to consistently refer to the CDDL concept? I'm confused what you want.

> A measurement-associated (sub-)environment is at the granularity of a single element, such as the second enumerated network interface card.

Reviewer 3: >I'm confused what you want.

I'm suggesting that before making changes to CDDL and spec text as it pertains to `mkey`, we should resolve the meta conversation of whether mkey is going to be removed. 
Issue #70 was tracking a use case that possibly justifies its use. The other use cases (that I'm aware of) are profile specific. 

There was a thread related to `mkey` as part of the internal representation conversation that resulted in the following CDDL being included:
```
ECT = {
  ? e: environment-map
  ? c: claims-map / [ + local-claim ]
  ? a: [ + $crypto-key-type-choice ]
  ? ns: text
  ? cm: cm-type
  ? p: $profile-type-choice
}
local-claim = {
  le: local-environment
  c: claims-map
}
local-environment =  bstr / tstr
```
The strongest reason to remove `mkey` is its another form of an ECT structure (that is awkwardly nested in the Claim part of a "parent?" ECT). Since we clarified that the identifier scope for environment values in `environment-map` is global. If an `mkey` has global scope, another ECT could be used instead.  

So, the salient benefit that `mkey` brings is some syntactic sugar when `mkey` identifiers are locally scoped (relative to the parent ECT) to make RIMs less verbose. Instead of having to define another OID or UUID to disambiguate multiple measurements of the same kind (i.e., multiple entries with the same measurement-values-map code point). A local scope identifier, such as a number or string, is used.

But we need to have this conversation in the context of a use case (e.g., issue #70). So, lets prioritize #70 above #248.

> The types of this measured element identifier are OID, UUID or uint.

Reviewer 3: Since there is a pending discussion around whether or not we want to key mkey. We should hold off on making changes to mkey related content until that discussion is resolved.
